in_toto/keylib.go

```diff
@@ -46,11 +78,147 @@ func ParseRSAPublicKeyFromPEM(pemBytes []byte) (*rsa.PublicKey, error) {
 }
 
 /*
-LoadPublicKey parses an RSA public key from a PEM formatted file at the passed
+GeneratePublicPemBlock creates a "PUBLIC KEY" PEM block from public key byte data.
+If successful it returns PEM block as []byte slice.
+*/
+func GeneratePublicPemBlock(pubKeyBytes []byte) []byte {
+	// construct PEM block
+	publicKeyPemBlock := &pem.Block{
+		Type:    "PUBLIC KEY",
+		Headers: nil,
+		Bytes:   pubKeyBytes,
+	}
+	return pem.EncodeToMemory(publicKeyPemBlock)
+}
+
+/*
+SetKeyComponents sets all components in our key object.
+Furthermore it makes sure to remove any trailing and leading whitespaces or newlines.
+*/
+func (k *Key) SetKeyComponents(pubKeyBytes []byte, privateKeyBytes []byte, keyType string, scheme string, keyIdHashAlgorithms []string) error {
+	if len(privateKeyBytes) > 0 {
+		// assume we have a privateKey
+		k.KeyVal = KeyVal{
+			Private: strings.TrimSpace(string(privateKeyBytes)),
+			Public:  strings.TrimSpace(string(GeneratePublicPemBlock(pubKeyBytes))),
+		}
+	} else {
+		k.KeyVal = KeyVal{
+			Public: strings.TrimSpace(string(pubKeyBytes)),
+		}
+	}
+	k.KeyType = keyType
+	k.Scheme = scheme
+	k.KeyIdHashAlgorithms = keyIdHashAlgorithms
+	if err := k.GenerateKeyId(); err != nil {
+		return err
+	}
+	return nil
+}
+
+/*
+LoadKey loads the key file at specified file path into the key object.
+It automatically derives the PEM type (PKCS1/PKCS8) and the key type (RSA/ed25519).
+On success it will return nil, otherwise an error.
+*/
+func (k *Key) LoadKey(path string, scheme string, keyIdHashAlgorithms []string) error {
+	pemFile, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if closeErr := pemFile.Close(); closeErr != nil {
+			err = closeErr
+		}
+	}()
+	// Read key bytes and decode PEM
+	pemBytes, err := ioutil.ReadAll(pemFile)
+	if err != nil {
+		return err
+	}
+
+	// TODO: There could be more key data in _, which we silently ignore here.
+	// Should we handle it / fail / say something about it?
+	data, _ := pem.Decode(pemBytes)
+	if data == nil {
+		return fmt.Errorf("could not find a private key PEM block")
+	}
+
+	// Try to load private key, if this fails try to load
+	// key as public key
+	key, err := x509.ParsePKCS8PrivateKey(data.Bytes)
+	if err.Error() == "x509: failed to parse private key (use ParsePKCS1PrivateKey instead for this key format)" {
```

Comment:
It's pretty convenient to fall back to PKCS1 here, but it also feels quite brittle to rely on the error message for case handling.

Moreover, looking at the sources of `x5009.ParsePKCS8PrivateKey` and `asn1.Unmarshal`, the suggestion in the error message seems more like a wild than an educated guess.

I'm fine with trying different parse functions, but I would just base it on the error and not on the error message. Btw. you could also try ParseECPrivateKey. :) What do you think?

Reply:
The common used `errors.Is(err, ErrPKCS1PrivKey)` does not work, because the library provides no such error. Therefore the only way to differ the error is via an error message. Is this correct and what you are asking for?

> I'm fine with trying different parse functions, but I would just base it on the error and not on the error message. Btw. you could also try ParseECPrivateKey. :) What do you think?

is EC == ECDSA? Isn't this covered with PKCS8?

Reply:
I think I got what you mean now. I just test on err != nil right now. This has made the whole section more readable, however I think I should make an own function out of this and maybe use a dispatch table for calling of these parsing functions in a row?

What do you think? If we keep adding parsing functions like ParseECPrivateKey we will have more and more nesting with the current approach, it looks horrible, but it's easy to read. A dispatch table would be more complex to read but it looks nicer? Not sure about this. What do you think @lukpueh?

Reply:
nevermind.. moved the part to a separate function.